lib/hanami/application/router.rb

```diff
@@ -1,74 +1,49 @@
 # frozen_string_literal: true
 
 require "hanami/router"
-require "rack"
+require "hanami/application/routing/middleware/stack"
 
 module Hanami
   class Application
     # Hanami application router
     # @since 2.0.0
-    class Router < Hanami::Router
-      def initialize(**options, &block)
-        @options = options
-        @middlewares = []
-        super
-        @rack_app = build_rack_app
-      end
-
-      def use(*args, &block)
-        @middlewares << (args << block)
+    class Router < ::Hanami::Router
+      # @since 2.0.0
+      # @api private
+      def initialize(stack:, **kwargs, &blk)
+        @stack = stack
+        super(**kwargs, &blk)
       end
 
-      # rubocop:disable Metrics/MethodLength
-      def mount(app, at:, host: nil, &block)
-        if app.is_a?(Symbol)
-          # TODO: I wonder if this is the actual behaviour we want...
-          raise "Slices can only be mounted from top-level of routes" unless @context.respond_to?(:slices)
-
-          # TODO: store slices in a way that's more efficient for lookup
-          slice = @context.slices[app]
+      # @since 2.0.0
+      # @api private
+      def freeze
+        return self if frozen?
 
-          # TODO: real exception class
-          raise "Slice +#{app}+ not found" unless slice
-
-          slice_router = self.class.new(
-            **@options,
-            context: slice,
-            endpoint_resolver: @endpoint_resolver.with_container(slice),
-            &block
-          )
-
-          super(slice_router, at: at, host: host)
-        else
-          super(app, at: at, host: host)
-        end
+        remove_instance_variable(:@stack)
```

Comment:
Would you mind explaining the reasoning for this? It's not immediately clear to me why we'd want to remove the stack when freezing...

Reply:
@timriley TL;DR: consider `Stack` as a builder for `Rack::Builder` that is not needed at the runtime.

`@stack` is only useful at "compile-time". It collects the Rack middleware usage from application configuration and `Router#use`, it applies the stack, but it isn't needed at the runtime.

When we invoke `Stack#finalize`, it returns the "real" middleware stack (a `Rack::Builder` instance). See https://github.com/hanami/hanami/pull/1032/files/d76a0cb0dd2b77dda5470e75adae3e837bda8e6f#diff-00f27089d3c7dbf5783b4a35405b603aR344

Reply:
Ahh, I see, thanks! I'll leave another comment over there.